### PR TITLE
broker: Unbind redis

### PR DESCRIPTION
**Pull request by indradhanush**
### Instructions for testing

Make sure you have a k8s cluster with the pre-requisites installed:
    - service catalog
    - habitat operator

(Read #7 for more)

```
make deploy-helm
svcat get brokers
```

Wait for the `habitat-service-broker` to be ready.

```
kubectl apply -f manifests/redis/service-instance.yaml
kubectl apply -f manifests/redis/binding.yaml
```

Wait for a new pod to be created.

Verify that secret was created:

```
kubectl get secrets | grep habitat-osb-redis
```

Verify that binding was created:

```
svcat get bindings
```

Check for `user.toml` file under `/hab/user/redis/config` in the pod:
```
kubectl exec redis-664f548f64-6hnhr  -ti bash
ls /hab/user/redis/config
```

**Note:** Exact name of the redis pod will be different than the example above.

Unbind the service instance:

```
svcat unbind redis-habitat
```

Verify that secret was deleted:

```
kubectl get secrets | grep habitat-osb-redis
```

Verify that binding was removed:

```
svcat get bindings
```

Check that `/hab/user` no longer exists in the pod:

```
kubectl exec redis--664f548f64-6hnhr  -ti bash
ls /hab
```

### CODE REVIEW

**Comment by iaguis** [comment]
I've fixed a couple of things in https://github.com/kinvolk/habitat-service-broker/commits/iaguis/unbind-redis, feel free to take them.

After those (and to my limited understanding) it seems to work.

File: cmd/servicebroker/main.go [hunk 1]
```diff
@@ -144,9 +145,15 @@ func setupKubeClient() (*broker.Client, error) {
...
+	if err != nil {
+		return nil, err
+	}
+
 	return &broker.Client{
-		KubeClient: apiclientset,
-		Client:     cl,
+		KubeClient:           apiclientset,
+		Client:               cl,
+		ServiceCatalogClient: svcatclientset,
```
Comment (iaguis) [inline cmd/servicebroker/main.go]: It seems this is not used?

File: pkg/broker/logic.go [hunk 2]
```diff
@@ -161,6 +164,24 @@ func generateHabitatObject(planID string) (*habv1beta1.Habitat, error) {
...
+		return nil, err
+	}
+
+	for i := range services.Items {
+		if services.Items[i].Spec.ExternalID == instanceID {
+			return &services.Items[i], nil
+		}
+	}
+
+	return nil, errors.New(fmt.Sprintf("No service instance found for instanceID: %q", instanceID))
```
[the remark this answers is not in the document]
Comment (indradhanush) [reply pkg/broker/logic.go]: Fixed.

File: cmd/servicebroker/main.go (hunk 1 above)
Comment (indradhanush) [reply cmd/servicebroker/main.go]: Yes it is being used in `getServiceInstance`: https://github.com/kinvolk/habitat-service-broker/pull/13/commits/af78566581bd4331f5eff52c75bf78ec0c1e64c1#diff-ba3324fac3ad609a2d797929efb07813R168

Comment (iaguis) [reply cmd/servicebroker/main.go]: ah, I got confused by those nested KubeClients

File: pkg/broker/logic.go [hunk 3]
```diff
@@ -242,6 +257,39 @@ func (b *BrokerLogic) createBinding(request *osb.BindRequest) error {
...
+	}
+
+	hab, err := b.GetHabitat(name)
+	if err != nil {
+		return fmt.Errorf("error getting Habitat service: %v", err)
+	}
+
+	switch name {
+	case "redis":
+		err := b.deleteSecret(hab.Spec.Service.ConfigSecretName)
```
Comment (krnowak) [inline pkg/broker/logic.go]: I would do it in reverse order compared to bind. When binding you create a secret and then update the habitat resource. So in unbind, you would need update the habitat resource first and then drop a secret.

File: pkg/broker/logic.go [hunk 4]
```diff
@@ -242,6 +257,39 @@ func (b *BrokerLogic) createBinding(request *osb.BindRequest) error {
...
+		if err != nil {
+			return fmt.Errorf("error updating habitat: %v", err)
+		}
+
+		err := b.deleteSecret(hab.Spec.Service.ConfigSecretName)
+		if err != nil {
+			return fmt.Errorf("error deleting secret: %v", err)
+		}
+	default:
+		return fmt.Errorf("binding for %q is not implemented.", name)
```
Comment (krnowak) [inline pkg/broker/logic.go]: unbinding

**Comment by indradhanush** [comment]
Merged on the command line.

**Pull request merged**